### PR TITLE
[FIRRTL][OM] Add property asserts

### PR DESCRIPTION
**Pull request by seldridge**
Add support for a FIRRTL property assert and an OM property assert.  This
is intended to allow for front-end language users to define metadata
constraints.  These constraints are either compile-time known to be
true/false or they are evaluated at metadata evaluation time.

This is part of building towards adding FIRRTL domain constraints.  At
present, FIRRTL only has maximally unsafe `firrtl.unsafe_domain_cast`
operations.  We'd like to have the ability to do this unsafe cast, but
apply additional constraints.  E.g., for a synchronous clock domain
crossing, we'd like to be able to say that the crossing must be between
two domains which are synchronous to each other.  However, at present we
cannot do this and a synchronous crossing looks the same as an
asynchronous crossing.  This is infrastructure to enable this kind of
check.

Commits are logical and should be reviewed independently.

AI-assisted-by: Claude Code (Claude Sonnet 4.6, Claude Haiku 4.5)

### CODE REVIEW

**Comment by fabianschuiki** [comment]
[Results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-01-031301-pr10093-7eb04bf) of circt-tests [run](https://github.com/circt/circt-tests/actions/runs/23829887030) for [7eb04bf](https://github.com/llvm/circt/commit/7eb04bf) compared to [results](https://github.com/circt/circt-tests/tree/results/2026/03/2026-03-31-225902-main-5dc62fe) for [5dc62fe](https://github.com/llvm/circt/commit/5dc62fe): no change to test results.

File: lib/Dialect/OM/Evaluator/Evaluator.cpp [hunk 1]
```diff
@@ -277,6 +277,28 @@ circt::om::Evaluator::evaluateObjectInstance(StringAttr className,
...
+  for (auto assertOp : cls.getOps<PropertyAssertOp>()) {
+    auto condResult =
+        evaluateValue(assertOp.getCondition(), actualParams, assertOp.getLoc());
+    if (failed(condResult))
+      return failure();
+    auto &condValue = *condResult;
+    // If the condition is unknown, skip silently (best-effort).
+    if (condValue->isUnknown())
+      continue;
+    if (auto *attrVal = dyn_cast<evaluator::AttributeValue>(condValue.get())) {
```
Comment (mikeurbach) [inline lib/Dialect/OM/Evaluator/Evaluator.cpp]: besides constant attributes, this could also be a field of an object. we have to handle something like that in the arithmetic ops for example: https://github.com/llvm/circt/blob/0fde1fe97d25fe69a20a17c45be1dfa435ae2aa6/lib/Dialect/OM/Evaluator/Evaluator.cpp#L461-L473

i would add some evaluator unit tests for this change in general. a couple edge cases we test for the arithmetic operations is a) the evaluator value is from object fields, and b) the evaluator value is from object fields, and there are cycles such that the value isn't ready right away.

specifically for that latter b) case, we have this isFullyEvaluated handling, which I don't see here yet: https://github.com/llvm/circt/blob/e712f4d44a171f0a3ff9bfb0ab1d5dedf77c75b4/lib/Dialect/OM/Evaluator/Evaluator.cpp#L441-L453

File: include/circt/Dialect/FIRRTL/FIRRTLStatements.td [hunk 2]
```diff
@@ -370,6 +370,34 @@ def PropAssignOp : FIRRTLOp<"propassign", [FConnectLike, SameTypeOperands,
...
+    Example:
+    ```mlir
+    firrtl.property_assert %cond, "invariant must hold" : !firrtl.bool
+    ```
+  }];
+
+  let arguments = (ins
+    BoolType:$condition,
+    StrAttr:$message,
+    OptionalAttr<StrAttr>:$name
```
Comment (uenoku) [inline include/circt/Dialect/FIRRTL/FIRRTLStatements.td]: Is a label needed even wen OM assert doesn't have one?

Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLStatements.td]: It's questionable. Probably not. This is now used in the lowering to the OM asserts, however, it's just used as a prefix if it exists.

File: lib/Dialect/OM/Evaluator/Evaluator.cpp (hunk 1 above)
Comment (seldridge) [reply lib/Dialect/OM/Evaluator/Evaluator.cpp]: Thanks for the solid suggestions here. I added much more exhaustive testing in the evaluator tests for these cases in 85918b618. WDYT?

Also, there is some duplication of testing across the (now more exhaustive) Google Test evaluator tests, Python tests (which I haven't run locally), and the CAPI evaluator tests. Does this split seem okay to you? (The tests that aren't in 85918b618 were AI-assisted... They look correct to me, but I'm not sure if they are important to be included or not.)

File: include/circt/Dialect/FIRRTL/FIRRTLStatements.td (hunk 2 above)
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLStatements.td]: Removed in later `!fixup` commits.

There's no real need for this. The FIRRTL assert/assume/cover label was motivated by a lowering to SystemVerilog. We don't need labels here in OM property asserts. We could have them, but they are no more special than just using a `<label>: ` prefix in the message.

Good idea.

**Comment by fabianschuiki** [comment]
[Results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-045103-pr10093-85918b6) of circt-tests [run](https://github.com/circt/circt-tests/actions/runs/24172246765) for [85918b6](https://github.com/llvm/circt/commit/85918b6) compared to [results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-004854-main-f5d48db) for [f5d48db](https://github.com/llvm/circt/commit/f5d48db): no change to test results.

**Comment by fabianschuiki** [comment]
[Results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-052431-pr10093-01b4cf8) of circt-tests [run](https://github.com/circt/circt-tests/actions/runs/24173688284) for [01b4cf8](https://github.com/llvm/circt/commit/01b4cf8) compared to [results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-045103-pr10093-85918b6) for [85918b6](https://github.com/llvm/circt/commit/85918b6): no change to test results.

File: lib/Dialect/OM/Evaluator/Evaluator.cpp [hunk 3]
```diff
@@ -512,6 +517,56 @@ circt::om::Evaluator::evaluateIntegerBinaryArithmetic(
...
+
+  auto condAttr = extractAttr(condResult.value().get());
+  if (!condAttr)
+    return success();
+
+  bool isFalse = false;
+  if (auto boolAttr = dyn_cast<BoolAttr>(condAttr))
+    isFalse = !boolAttr.getValue();
+  else if (auto intAttr = dyn_cast<mlir::IntegerAttr>(condAttr))
+    isFalse = intAttr.getValue().isZero();
```
Comment (uenoku) [inline lib/Dialect/OM/Evaluator/Evaluator.cpp:563]: Does the ill-typed condition silently pass asserts? Or is the type checked somewhere else?

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 4]
```diff
@@ -4410,6 +4410,17 @@ LogicalResult PropAssignOp::verify() {
...
 
+LogicalResult PropertyAssertOp::verify() {
+  // Static evaluation: if the condition is a known constant false, the
+  // assertion is trivially violated and we can report an error immediately.
+  if (auto *defOp = getCondition().getDefiningOp())
+    if (auto boolConst = dyn_cast<BoolConstantOp>(defOp))
+      if (!boolConst.getValue())
+        return emitOpError("property assertion is statically false: ")
+               << getMessage();
+  return success();
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4421]: I'm fine to reject in verifier but I think assertion failure is fine if it's not reachable from the entry point.

Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:4421]: The case that I'm trying to catch with this is situations like a too-weak clock domain synchronizer. E.g., there's roughly three kinds of synchronizers: (1) asynchronous, (2) rational, and (3) synchronous. (1) is the strongest and can handle any two clock domains, (2) can handle any clocks derived from the same source with a rational relationship, and (3) is just a direct connection, but needs the clocks to be a power-of-two multiple  so that the clock edges are always aligned.

I'm trying to provide a "type error" if (3) is used in place of (1) and am worried that if we wait until SDC generation, that this will be too late to be useful.

Can you think of any other "error early" approaches?

File: lib/Dialect/OM/Evaluator/Evaluator.cpp (hunk 3 above)
Comment (seldridge) [reply lib/Dialect/OM/Evaluator/Evaluator.cpp:563]: Good catch. Yeah, this could leak through with the default. I'm not sure if this is guaranteed to be one of these, so I added a check in: f93ee5dd3

**Comment by fabianschuiki** [comment]
[Results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-164923-pr10093-f93ee5d) of circt-tests [run](https://github.com/circt/circt-tests/actions/runs/24201903784) for [f93ee5d](https://github.com/llvm/circt/commit/f93ee5d) compared to [results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-052431-pr10093-01b4cf8) for [01b4cf8](https://github.com/llvm/circt/commit/01b4cf8): no change to test results.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 4 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:4421]: > (3) is just a direct connection, but needs the clocks to be a power-of-two multiple so that the clock edges are always aligned.

I feel this requires actual elaboration in many cases as I can imagine these parameters are derived from the top level module, e.g.:
```
assert clock_a == clock_b * 2
```
clock_a and clock_b are certainly not constant locally. 
 
That said I'm not opposed to this verifier.

Comment (seldridge) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:4421]: I've been trying to avoid encoding actual periods as it's rare that a clock is ever actually fixed, e.g., consider DVFS. The modeling I'd like to work towards is:

```
class RelatedClock:
  sourceClock: Inst<ClockDomain>
  relationship: Rational | Synchronous

domain ClockDomain:
  name: String
  relatedClock: Option<Inst<RelatedClock>>
```

Then the check can be tightly written such that it's just checking that it has the correct related clock and relationship.

Note: I'm hacking and slashing my way towards this, but using a weaker version with strings, booleans, and integers which is hidden via Chisel APIs. However, this is the direction I'd like to take Domains (Properties).

**Comment by fabianschuiki** [comment]
[Results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-224759-pr10093-85185c4) of circt-tests [run](https://github.com/circt/circt-tests/actions/runs/24216751022) for [85185c4](https://github.com/llvm/circt/commit/85185c4) compared to [results](https://github.com/circt/circt-tests/tree/results/2026/04/2026-04-09-004854-main-f5d48db) for [f5d48db](https://github.com/llvm/circt/commit/f5d48db):

### sv-tests

Changes in emitted diagnostics:
- **-9 total change**
- -4 `` error: unsupported system call `$fopen` ``
- -2 `error: endianness of selection must match declared range (type is 'logic[21:0]')`
- -2 `error: unknown module 'BUFG'`
- +2 `error: identifier 'in' used before its declaration`
- -1 `error: unknown module 'IBUF'`
- -1 `error: unknown module 'PLLE2_ADV'`
- -1 `error: variable 'regs68H' driven by always_ff procedure cannot be written to by any other process`
- -1 `error: variable 'regs68L' driven by always_ff procedure cannot be written to by any other process`
- +1 `error: unknown module 'BUFGCE'`
